### PR TITLE
fix(npm): better node version and version requirement compatibility

### LINKED ISSUE #15600: npm: error parsing version requirement for dependency
## Steps to Reproduce
```
$ deno --version
deno 1.25.0 (release, aarch64-apple-darwin)
v8 10.6.194.5
typescript 4.7.4

$ cat main.ts
import { scaleLinear } from "npm:d3-scale@4.0.2";

$ deno run --unstable main.ts
error: npm package: d3-scale@4.0.2

Caused by:
    0: error parsing version requirement for dependency: d3-time@2.1.1 - 3
    1: expected comma after patch version number, found '-'
```

For reference https://github.com/d3/d3-scale/blob/main/package.json#L35-L40

### PR DESCRIPTION
**Pull request by dsherret**
We initially used the `semver` crate as a first pass solution, but this PR adds way better node semver support.

Wrote an in-house solution so we can move fast fixing bugs and re-use some of the code for our npm specifiers. The npm specifiers use a stricter parser than the version requirements allowed in package.json files.

Closes #15600

### CODE REVIEW

File: cli/Cargo.toml [hunk 1]
```diff
@@ -78,7 +78,8 @@ indexmap = "1.8.1"
 jsonc-parser = { version = "=0.21.0", features = ["serde"] }
 libc = "=0.2.126"
 log = { version = "=0.4.17", features = ["serde"] }
-mitata = '=0.0.7'
+mitata = "=0.0.7"
+monch = "=0.2.0"
```
Comment (dsherret) [inline cli/Cargo.toml:82]: Already a dependency of deno_task_shell.

File: cli/npm/semver/mod.rs [hunk 2]
```diff
@@ -0,0 +1,997 @@
...
+      ("1.0.0", "1.0.0"),
+      (">=*", "0.2.4"),
+      ("", "1.0.0"),
+      // ("*", "1.2.3", {}),
+      // ("*", "v1.2.3", { loose: 123 }),
+      // (">=1.0.0", "1.0.0", /asdf/),
+      // (">=1.0.0", "1.0.1", { loose: null }),
+      // (">=1.0.0", "1.1.0", { loose: 0 }),
+      // (">1.0.0", "1.0.1", { loose: undefined }),
+      (">1.0.0", "1.1.0"),
```
Comment (bartlomieju) [inline cli/npm/semver/mod.rs:769]: Why are some of these commented out?

File: cli/npm/semver/specifier.rs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    assert!(!tester.matches("1.0.0"));
+
+    let tester = VersionReqTester::new("~0.0.1");
+    assert!(!tester.matches("0.0.0"));
+    assert!(tester.matches("0.0.1"));
+    assert!(tester.matches("0.0.2")); // for some reason this matches, but not with ^
+    assert!(!tester.matches("0.1.0"));
+    assert!(!tester.matches("1.0.0"));
+  }
+}
```
Comment (bartlomieju) [inline cli/npm/semver/specifier.rs:250]: Nice tests!

**Review by bartlomieju — APPROVED** [review]
LGTM!

File: cli/npm/semver/mod.rs (hunk 2 above)
Comment (dsherret) [reply cli/npm/semver/mod.rs]: I forgot to investigate these ones. Just did now and made the parsing looser.

**Pull request merged**